### PR TITLE
plan/typeinferer: refine type inferer of builtin func greatest/least

### PR DESCRIPTION
**Pull request by XuHuaiyu**
refine the type inferer of greatest() and least().

PTAL @coocood @hanfei1991 @zimulala @tiancaiamao

### CODE REVIEW

File: plan/typeinferer.go [hunk 1]
```diff
@@ -221,6 +221,45 @@ func mergeArithType(fta, ftb *types.FieldType) byte {
...
+		ft.Tp = mysql.TypeNewDecimal
+	} else if fta.Tp == mysql.TypeLonglong || ftb.Tp == mysql.TypeLonglong {
+		ft.Tp = mysql.TypeLonglong
+	} else {
+		ft.Tp = mysql.TypeLong
+	}
+	return ft
+}
+
+func isBinary(ft *types.FieldType) bool {
```
Comment (shenli) [inline plan/typeinferer.go]: You can use types.HasBinaryFlag().

File: plan/typeinferer.go [hunk 2]
```diff
@@ -221,6 +221,45 @@ func mergeArithType(fta, ftb *types.FieldType) byte {
 	return mysql.TypeLonglong
 }
 
+func mergeCmpType(fta, ftb *types.FieldType) (ft *types.FieldType) {
+	ft = &types.FieldType{}
+	if fta.Charset == charset.CharsetUTF8 && ftb.Charset == charset.CharsetUTF8 {
+		ft.Charset = charset.CharsetUTF8
+		ft.Collate = "utf8_bin"
```
Comment (shenli) [inline plan/typeinferer.go]: Should we set ft.Charset/Collate in other cases?

**Comment by shenli** [comment]
@XuHuaiyu Any update?

File: plan/typeinferer.go (hunk 2 above)
Comment (XuHuaiyu) [inline plan/typeinferer.go]: In other cases, ft.Charset/Collate will be set as CharsetBin/CollateBin outside this func.

**Comment by XuHuaiyu** [comment]
PTAL
@shenli @coocood

File: plan/typeinferer.go (hunk 2 above)
Comment (coocood) [inline plan/typeinferer.go]: use constant `mysql.UTF8DefaultCollation`

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**